### ISSUE TITLE
New lint: wildcard pattern in let

### ISSUE THREAD

**Issue by Vlad-Shcherbina** [rust-lang/rust-clippy#4090]
I think there should be a warning about the wildcard pattern in let.
I can't imagine any situation where this makes sense, and there are cases where it's a mistake caused by taking `_` for identifier though it's not.

Wrong:
```rust
{
    let _ = my scope guard or whatever
    // dropped right away

    do stuff
}
```

Correct:
```rust
{
    let _g = my scope guard or whatever

    do stuff

    // dropped at the end of the scope
}

**Comment by flip1995** [comment]
One use case is to prevent warnings on `#[must_use]` functions.
```rust
let _ = returning_result_i_dont_want_to_deal_with()
```

But this could be implemented as an restriction lint.

**Comment by Manishearth** [comment]
We could specifically lint on known guard types

**Comment by Vlad-Shcherbina** [comment]
Many known guards (like `std::cell::Ref`, `std::sync::MutexGuard`) are meant to be accessed in scope, so this warning would be unlikely to trigger.
And I ran into this problem with a custom guard type in the first place.

Indeed I forgot that `let _ =` is idiomatically used to suppress `must_use` warnings.
I wish there was `unused()` function in the standard library (same behavior as `std::mem::drop()`) instead.

-------------

Two ideas that don't work:
1. Warn iff there is custom Drop impl. The problem is that it will cause false positives on functions returning strings.
2. Warn iff the expression isn't annotated with `must_use`. The problem is that properly implemented guards are likely `must_use`.

**Comment by lolbinarycat** [comment]
i feel like this makes a decent restriction lint.  `let _` is fine for most codebases, but some may wish to actually use all must_use types, and this enables that (along with `forbid(unused_must_use)`)

**Comment by flip1995** [comment]
> and this enables that (along with `forbid(unused_must_use)`)

`unused_must_use` doesn't trigger on `let _ = must_use();`: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21a92caae7604dfd0d73b65beabd5b73). So this lint would be a prior step to get this lint to trigger.

**Comment by lolbinarycat** [comment]
yes, that was my point.  this lint (once implemented) would be mainly useful in conjunction with `unused_must_use`.

i have [a partial implementation of this](https://github.com/lolbinarycat/rust-clippy/tree/wildcard_let), and i successfully registered the lint.. but for some reason it never produces any output, not even debug prints?

**Comment by flip1995** [comment]
To add a new lint, using [`cargo dev new_lint`](https://doc.rust-lang.org/clippy/development/defining_lints.html?highlight=register#add-and-register-the-lint) is usually the easiest way. Now that you have down that manually, you still have to [register the lint](https://doc.rust-lang.org/clippy/development/defining_lints.html?highlight=register#lint-registration) (bottom of the page)

**Comment by lolbinarycat** [comment]
personally I think this already covered by the [let_underscore_must_use](https://rust-lang.github.io/rust-clippy/master/index.html#/let_underscore_must_use) lint

@rustbot release-assignment
@rustbot label -good-first-issue
